### ISSUE TITLE
Support for Service Chaining

### ISSUE THREAD

**Issue by t-lin** [Multi-Tier-Cloud/service-manager#50]
Need to design protocol and figure out how to forward traffic through a series of proxies.

May also need to consider how to return the traffic:
- Option 1: Return directly from destination service's proxy back to orignal client's proxy
- Option 2: Return along the same path (e.g. traverse VNFs in reverse-order)
- Option 3: Return along the same set of VNFs (e.g. traverse VNFs in the same order)

Option 1 is the easiest to do, so probably will support that. Once that exists, supportion options 2 and 3 should be almost identical to option 1 (i.e. destination simply specifies chain of services).

Important considerations:
- At the client proxy, it must be able to handle the returned data (which may arrive on a separate P2P stream from another peer) and return it on the original HTTP, TCP, or UDP connection from the proxy.

Looking into https://github.com/libp2p/go-libp2p-circuit for implementation ideas.

**Comment by t-lin** [comment]
Possible spec for HTTP proxy (using simple GET) that supports basic arguments:
- `http://clientProxy/service1/service2/service3?service1=s1arg1,s1arg2,s1arg3&service3=s3arg1`
- The above example could be translated into the following queries to the actual services:
  - `http://service1/s1arg1/s1arg2/s1arg3`
  - `http://service2/`
  - `http://service3/s3arg1`

Possible spec for L4 proxy:
- `http://clientProxy/tcp/service1/service2/service3`
  - This forms a TCP chain from client to service1, to service2, and to service3
- `http://clientProxy/tcp/service1/service2/udp/service3`
  - This forms a TCP chain from client to service1 and to service2, then from there, a UDP connection to service3

**Comment by hivanco** [comment]
I'm thinking the arguments should be handled a bit differently. When querying the actual service you you might want both a subpath and query parameters following the service address. Eg. `http://service1/s1path?s1arg=s1value`.

So for HTTP proxy spec, at least one possibility I see:
- service1 query parameter represents the string literal that should follow service1 in the actual query
  - `http://clientProxy/service1/service2?service1=/s1path?s1arg=s1value`
  - Would need to encode the `/, ?, =` in the query string of course

**Comment by t-lin** [comment]
Yeah, that's a good point. Another alternative I just realized... we could embed the parameters within the HTTP headers, since it allows arbitrary key-values.
